### PR TITLE
add datacopy contract test

### PR DESCRIPTION
**Pull request by Liuhaai**
verified the fix for CVE-2021-39137 on our chain

### CODE REVIEW

File: action/protocol/execution/testdata-istanbul/datacopy.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
+pragma solidity ^0.8.3;
+
+contract Datacopy {
+    function dataCopy() public view returns (bool) {
+        bytes memory arr = new bytes(3);
+        arr[0] = 0x11;
+        arr[1] = 0x22;
+        arr[2] = 0x00;
```
Comment (dustinxie) [inline action/protocol/execution/testdata-istanbul/datacopy.sol:8]: with the fix, we should have [0x11, 0x22, 0x00] after the copy, can we check that?
like `return arr[0] = 0x11 && arr[1] = 0x22 && arr[2] = 0x00`

File: action/protocol/execution/testdata/datacopy.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        "rawGasLimit": 5000000,
+        "rawGasPrice": "0",
+        "rawExpectedGasConsumed": 222483,
+        "expectedBalances": [],
+        "comment": "deploy datacopy contract"
+    }],
+    "executions": [{
+        "readOnly": true,
+        "rawPrivateKey": "cfa6ef757dee2e50351620dca002d32b9c090cfda55fb81f37f1d26b273743f1",
+        "rawByteCode": "574c8078",
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.json:19]: can we construct a raw bytes similar to attack tx: https://tx.blocksecteam.com/tx/0x1cb6fb36633d270edefc04d048145b4298e67b8aa82a9e5ec4aa1435dd770ce4
and play the tx on testnet?

File: action/protocol/execution/testdata-istanbul/datacopy.sol (hunk 1 above)
Comment (Liuhaai) [reply action/protocol/execution/testdata-istanbul/datacopy.sol:8]: Unable to catch the return value unless the code in core/vm/evm.go is modified

**Review by raullenchai — APPROVED** [review]

File: action/protocol/execution/testdata/datacopy.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        "rawAmount": "0",
+        "rawGasLimit": 1000000,
+        "rawGasPrice": "0",
+        "rawExpectedGasConsumed": 12481,
+        "hasReturnValue": true,
+        "rawReturnValue": "000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000031122330000000000000000000000000000000000000000000000000000000000",
+        "comment": "the data of return is [0x11, 0x22, 0x33]"
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.json:26]: as we discussed, use fixed privatekey and timestamp, add the block hash/state root/receipt root values into test?

File: action/protocol/execution/testdata-istanbul/datacopy.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        "rawByteCode": "574c8078",
+        "rawAmount": "0",
+        "rawGasLimit": 1000000,
+        "rawGasPrice": "0",
+        "rawExpectedGasConsumed": 12481,
+        "hasReturnValue": true,
+        "rawReturnValue": "000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000031122330000000000000000000000000000000000000000000000000000000000",
+        "comment": "the data of return is [0x11, 0x22, 0x33]"
+    }]
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata-istanbul/datacopy.json:28]: `make fmt`

File: action/protocol/execution/testdata/datacopy.sol [hunk 5]
```diff
@@ -0,0 +1,20 @@
+pragma solidity ^0.8.3;
+
+contract Datacopy {
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.sol:3]: bytes[40] arr;

function set() {
     arr[0] = 0x11;
     arr[1] = 0x22;
     arr[2] = 0x33;
}

File: action/protocol/execution/testdata-istanbul/datacopy.sol [hunk 6]
```diff
@@ -0,0 +1,20 @@
+pragma solidity ^0.8.3;
+
+contract Datacopy {
```
Comment (dustinxie) [inline action/protocol/execution/testdata-istanbul/datacopy.sol:3]: bytes[3] storage

File: action/protocol/execution/testdata-istanbul/datacopy.sol [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        arr[2] = 0x33;
+        uint length = arr.length;
+        bytes memory result = new bytes(length);
+        bool ret;
+        assembly {
+            // Call precompiled contract to copy data
+            ret :=staticcall(0x10000, 0x04, add(arr, 0x20), length, add(arr, 0x21), length)
+            // copy returnData into result
+            returndatacopy(add(result, 0x20), 0x00, length)
+        }
```
Comment (dustinxie) [inline action/protocol/execution/testdata-istanbul/datacopy.sol:17]: storage[0] = result[0]
storage[1] = result[1]
storage[2] = result[2]

File: action/protocol/execution/testdata-istanbul/datacopy.sol [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        return result;
+    }
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata-istanbul/datacopy.sol:20]: add a new line

File: action/protocol/execution/testdata/CVE-2021-39137-attack-replay.json [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        "rawGasPrice": "0",
+        "rawExpectedGasConsumed": 12300,
+        "comment": "launch attack(https://etherscan.io/tx/0x1cb6fb36633d270edefc04d048145b4298e67b8aa82a9e5ec4aa1435dd770ce4)",
+        "expectedBlockInfos" : {
+            "txRootHash" : "a59ead74a3870e9b5ca352c5f59108df402ca203ef2109799fe2d8e1da49c83d",
+            "stateRootHash" : "ed9bd589ee5ab5660a3d5d863bbeea13020a0aacab18e8655a626beaf9a54713",
+            "receiptRootHash" : "3285579efa8521fbf95829b868ff5d37632c4feac6167e9ab2dc4961004c9272"
+        }
+    }]
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata/CVE-2021-39137-attack-replay.json:30]: add a new line

File: action/protocol/execution/testdata/datacopy.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        "rawGasPrice": "0",
+        "rawExpectedGasConsumed": 32578,
+        "comment": "the data of return is [0x11, 0x22, 0x33]",
+        "expectedBlockInfos" : {
+            "txRootHash" : "3672945ca662bea4dd977e799374c5ce36c0c3e9ecbe98f1655f33439bbfe40c",
+            "stateRootHash" : "e55ecb27070c4e3eb3ae270a28caead35e9442b279781d6bb286434e526f2a29",
+            "receiptRootHash" : "d0891f793a8762e15b870357701eaddf56ce0177f7e26ddeeb5689e1ad36a1ab"
+        }
+    }]
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.json:30]: add a new line

File: action/protocol/execution/testdata/datacopy.sol [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    
+    function updateStore(bytes memory ret) internal {
+        store = ret;
+    }
+    
+    function printStore() public view returns (bytes memory){
+        return store;
+    }
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.sol:28]: add a new line

File: action/protocol/execution/testdata/datacopy.sol [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        arr[2] = 0x33;
+        uint length = arr.length;
+        bytes memory result = new bytes(3);
+        bool ret;
+        assembly {
+            // Call precompiled contract to copy data
+            ret := staticcall(0x10000, 0x04, add(arr, 0x20), length, add(arr, 0x21), length)
+            returndatacopy(add(result, 0x20), 0x00, length)
+        }
+        updateStore(result);
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.sol:18]: just `store = result` here

File: action/protocol/execution/testdata/datacopy.sol [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    
+    function updateStore(bytes memory ret) internal {
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.sol:21]: no need for this

File: action/protocol/execution/testdata/datacopy.sol [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        store = ret;
+    }
+    
+    function printStore() public view returns (bytes memory){
```
Comment (dustinxie) [inline action/protocol/execution/testdata/datacopy.sol:25]: no need for this?

File: action/protocol/execution/protocol_test.go [hunk 15]
```diff
@@ -258,7 +267,7 @@ func runExecutions(
 	ap actpool.ActPool,
 	ecfgs []*ExecutionConfig,
 	contractAddrs []string,
-) ([]*action.Receipt, error) {
+) ([]*action.Receipt, error, *ExpectedBlockInfo) {
```
Comment (dustinxie) [inline action/protocol/execution/protocol_test.go:270]: put `error` as last return value

**Review by dustinxie — APPROVED** [review]

**Pull request merged**